### ISSUE TITLE
Unable to connect to endpoint

### ISSUE THREAD

**Issue by vigneshprajapati** [awslabs/amazon-kinesis-producer#66]
While running the following command : 
mvn exec:java -Dexec.mainClass="com.amazonaws.services.kinesis.producer.sample.SampleProducer"

I am getting the below given error message: 
[kpl-daemon-0003] ERROR com.amazonaws.services.kinesis.producer.LogInputStreamReader - [2016-10-03 19:23:14.030505] [0x000018c2][0x00007fa27dffb700] [error] [shard_map.cc:150] Shard map update for stream "producer_test" failed. Code:  Message: Unable to connect to endpoint; retrying in 3375 ms

I have tried setting environment variables as per instruction provided at https://github.com/awslabs/amazon-kinesis-producer/tree/master/java/amazon-kinesis-producer-sample. But still this code doesnt able to connect with Kinesis stream. 

Can you suggest fix for the above ?

[When I try AWS CLI command: aws kinesis list-streams, its working. ]

**Comment by pfifer** [comment]
Are you behind a proxy.  I don't believe the KPL supports proxies right now.  Also have you created the stream for the SampleProducer?

**Comment by vigneshprajapati** [comment]
No, I am not using proxy. Also I have already created steam for SampleProducer. Still not ble to connect. !!

**Comment by pfifer** [comment]
Some additional questions to help:
* Are you using the default region of the sample, or a different region?
* Are you running the sample from an EC2 instance, or somewhere else?

**Comment by vigneshprajapati** [comment]
@pfifer 

1. Are you using the default region of the sample, or a different region?
    No. I am using 'us-east-1' region.
2. Are you running the sample from an EC2 instance, or somewhere else?
    No, I am trying it on local machine.

**Comment by pfifer** [comment]
Can you try adding this to ProducerSample.java to verify the connectivity to Kinesis?  This doesn't use exactly the same connectivity as the native components, but should at least help verify the case.

``` java
  public static void main(String[] args) throws Exception {
        //
        // Run some checks to make sure everything is good before starting the demo application. Even in production
        // application having some of these check can be useful as sanity tests.
        //
        preflightChecks();

       // Rest of the main method
}
```

``` java
    private static void preflightChecks() {
        Region region = Region.getRegion(Regions.fromName(REGION));

        AmazonKinesis kinesisClient = new AmazonKinesisClient(new DefaultAWSCredentialsProviderChain())
                .withRegion(region);

        ListStreamsResult listStreamsResult;
        try {
            listStreamsResult = kinesisClient.listStreams();
        } catch (AmazonServiceException ase) {
            errorMessage("Exception while list streams");
            errorMessage("-----------------------------");
            ase.printStackTrace(System.err);
            errorMessage("-----------------------------");
            errorMessage("Failed to list streams due to '%s'");
            errorMessage("Unable to list your streams in %s", REGION);
            String kinesisEndpoint = region.getServiceEndpoint("kinesis");
            errorMessage("Please make sure that credentials are available to the demo, "
                    + "and you have connectivity to reach %s", kinesisEndpoint);

            throw new RuntimeException("Can't connect to Kinesis to retrieve stream names", ase);
        }

        Set<String> streamNames = new HashSet<>();
        streamNames.addAll(listStreamsResult.getStreamNames());

        if (!streamNames.contains(STREAM_NAME)) {
            errorMessage("No stream names '%s' exists in region '%s'.", STREAM_NAME, region.getName());
            errorMessage("Was the stream created in another region?");

            throw new RuntimeException("Stream '" + STREAM_NAME + "' doesn't exist");
        }

        DescribeStreamResult describeStreamResult;
        try {
            describeStreamResult = kinesisClient.describeStream(STREAM_NAME);
        } catch (ResourceNotFoundException rnfe) {
            errorMessage("Exception while trying to describe stream '%s' in region '%s'.", STREAM_NAME,
                    region.getName());
            errorMessage("-----------------------------");
            rnfe.printStackTrace(System.err);
            errorMessage("-----------------------------");
            errorMessage("Stream '%s' was removed in between calls");

            throw new RuntimeException(rnfe);
        } catch (AmazonServiceException ase) {
            errorMessage("Exception while trying to describe stream '%s' in region '%s'.", STREAM_NAME,
                    region.getName());
            errorMessage("-----------------------------");
            ase.printStackTrace(System.err);
            errorMessage("-----------------------------");
            errorMessage(
                    "It's possible that the provided credentials don't have access rights to call describe stream.");

            throw new RuntimeException(ase);
        }
        StreamStatus status = StreamStatus.fromValue(describeStreamResult.getStreamDescription().getStreamStatus());
        if (status != StreamStatus.ACTIVE) {
            errorMessage("The stream '%s' exists, but isn't active.  "
                    + "Please wait for the stream to become active and try again.", STREAM_NAME);

            throw new RuntimeException("Stream not active.");
        }

    }
```

**Comment by eboto** [comment]
I am seeing this exact same behavior, but with these deviations:
  * Executing from within ec2
  * Region of `us-west-2`

I have executed your preflight code and get no output.

**Comment by eboto** [comment]
OK after characterizing it a bit further, this is either an SSL cert or file access issue. I've attached [some mildly edited logs](https://github.com/awslabs/amazon-kinesis-producer/files/517256/kpl_ssl_errors.txt) to show the point. The KPL does the following:

  * Set the CAs to `/tmp/amazon-kinesis-producer-native-binaries`
  * Attempt to use cURL, supporting just those CAs.
  * Fail to connect with either the Amazon or Cloudwatch endpoint with Curl error 77
    - According to [the curl docs](https://curl.haxx.se/libcurl/c/libcurl-errors.html), that error is `CURLE_SSL_CACERT_BADFILE. Problem with reading the SSL CA cert (path? access rights?)`

The weird thing is...it should have no problem finding or using the file. The file's permissions are accessible: `-rw-r--r-- 1 dfxAdmin     2563 Oct  7 22:59 b204d74a.0`

Its contents are exactly the same as [in the repo](https://github.com/awslabs/amazon-kinesis-producer/blob/master/java/amazon-kinesis-producer/src/main/resources/cacerts/b204d74a.0)

Any thoughts @pfifer ?

**Comment by artyomboyko** [comment]
@eboto I experience exactly the same problem. Also using docker.

**Comment by pfifer** [comment]
I'm looking into this.  The creation of the process on the Java side does set the CA directory.

The change from 0.10.1 to 0.12.0 moved from a custom HTTP/TLS client to the AWS SDK, and Curl as the client.  This didn't come up in my testing, so I need to try and track it down.  I was starting some verification for ECS for another issue, so I may head down that path to see I can replicate it.

**Comment by artyomboyko** [comment]
@pfifer 
Let me know if I can help. I can provide simple project (bin/conf/lib/share folders) with basic Dockerfile.

**Comment by artyomboyko** [comment]
@pfifer 

any advice on how I can see logging from CPP side of things? I'm setting default logging level is `info`, but I don't see output messages like this https://github.com/awslabs/amazon-kinesis-producer/blob/master/aws/kinesis/main.cc#L345 in my console

**Comment by pfifer** [comment]
I've been able to see that logging message at startup for my tests.  Let me check a simple example to see if I can re-verify it.

**Comment by pfifer** [comment]
There is a race condition on emitting that message.  I thought there might be, but I figured it was in a different place.

After creating the KinesisProducer, add a sleep.  I found about 500 milliseconds gives the native process time to startup.  Really the KinesisProducer probably needs a start operation, that blocks until the underlying process is ready to handle messages.

**Comment by pfifer** [comment]
Was able to recreate this running in Docker.  Still need to investigate it some more.

**Comment by isac** [comment]
If it helps, I got the same problem when running from my virtual machine (Debian) using VirtualBox.
Version 0.10.2 works fine.

**Comment by pfifer** [comment]
So at least have an idea of what is going on.  On my test system while executing in Docker, an extra certificate file parameter is getting sent to OpenSSL.  The file doesn't exist which causes OpenSSL to return before attempting to apply the directory based certificate look ups.

**Comment by pfifer** [comment]
I've found the issue, and am preparing a fix for it now.

**Comment by pfifer** [comment]
The PR is out as #68 

I'm working on getting everything ready for release.

**Comment by pfifer** [comment]
I've release 0.12.1 that should fix this.  Let me know if anyone is still having problems.

**Comment by pfifer** [comment]
This has been released.  Please reopen if you have any further issues.

**Comment by linevans** [comment]
I'm still seeing this error with version 0.12.3

I use a delay of 60 seconds after creating the KinesisProducer before I start using it to add records. I can work-around the issue by setting the KinesisConfiguration: setVerifyCertificate(false)

**Comment by clintmiller1** [comment]
I don't know if we're seeing the same issue that others in this thread are seeing or not. But in our production system, we intermittently get 2-3 "Unable to connect to endpoint" errors per day. That's in a Kinesis stream that's processing maybe 50k records per day. So the error rate is is low but annoying. We're running 0.12.3. Our producers are all running as Docker containers in EC2. Everything is in us-east-1a.

Our producers all launch a KPL instance when they startup and leave it running indefinitely, periodically using it to publish records. We are not seeing this error at startup time. So it's not a matter of us needing to sleep for a few seconds after creating the KPL. The error happens intermittently after we have published many records.

I'm reluctant to do setVerifyCertificate(false). That sounds like a security issue.

**Comment by pradeepcheers** [comment]
Same issue for me too. I'm using 0.12.3 amazon-kinesis-producer.

**Comment by Shilpa-halegoudru** [comment]
Same issue, I am using v0.12.5.

**Comment by zodeus** [comment]
I'm seeing this, trying to connect to localstack/localstack docker container 

`ERROR LogInputStreamReader - [2017-08-16 23:26:35.894466] [0x00000076][0x00007f3cecc7d700] [error] [shard_map.cc:150] Shard map update for stream "xxx" failed. Code:  Message: Unable to connect to endpoint; retrying in 1000 ms
 ERROR LogInputStreamReader - [2017-08-16 23:26:35.989101] [0x00000076][0x00007f3cecc7d700] [error] [AWS Log: ERROR](CurlHttpClient)Curl returned error code 35`

**Comment by etspaceman** [comment]
@zodeus I was getting the same issue, but realized that I had to start localstack with USE_SSL=true so that the KPL could connect properly (as the KPL always uses SSL). You can use the same variable in their awslocal-cli for options like create-stream.

**Comment by aecc** [comment]
@zodeus @etspaceman 
At the same time if you wanna access kinesis **localstack** using AWS java sdk you need to set these configs:
```
System.setProperty(AWS_CBOR_DISABLE_SYSTEM_PROPERTY, "true");
System.setProperty(DISABLE_CERT_CHECKING_SYSTEM_PROPERTY, "");
```

**Comment by auwing** [comment]
Hi All, 
     Don't know has anyone get this issue resolved? If yes, does it require any specify setup? I tried with aecc's solution, but had no success. Could someone please shed some light?

**Comment by twnbay78** [comment]
I am getting the same exact issue. This may be **entirely** unrelated, but I am getting an SSL issue when issuing AWS CLI commands. 

$aws kinesis list-streams --region us-east-1 

SSL validation failed for https://kinesis.us-east-1.amazonaws.com/ [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate in certificate chain (_ssl.c:1056)

I am behind a proxy and configured the proxy within my IDE (Intellij). After a test connection to any endpoint, I always get an "SSL Cert. not trusted" message. It would not be wise to disable ssl verification. I have tried to manually add amazon certs from here: https://www.amazontrust.com/repository/ 
to no avail. Anyone who can assist here would be greatly appreciated.
